src/models/models.jl

```diff
@@ -110,22 +116,113 @@ function InfiniteMPOMatrix(model::Model, s::CelledVector, translator::Function;
       if isnothing(idx)
         Hmat[range_H + 1 - n, range_H - n] = identity
       else
-        Hmat[range_H + 1 - n, range_H - n] = temp_H[j - n][idx]#replacetags(linkinds, temp_H[j - n][idx], "Link, l=$n", tags(ls[j-1]))
+        #Here, we split the local tensor into its different blocks
+        T = eltype(temp_H[j - n][idx])
+        temp_mat = convert_itensor_to_itensormatrix(
+          temp_H[j - n][idx];
+          leftdir=ITensors.In,
+          first=n == 0 ? true : false,
+          last=n == range_H - 1 ? true : false,
+          left_tags=tags(ls[j - 1]),
+          right_tags=tags(ls[j]),
+          leftindex=if idx > 1
+            only(commoninds(temp_H[j - n][idx], temp_H[j - n][idx - 1]))
+          else
+            nothing
+          end,
+        )
+        if size(temp_mat) == (3, 3)
+          @assert iszero(temp_mat[1, 2])
+          @assert iszero(temp_mat[1, 3])
+          @assert iszero(temp_mat[2, 3])
+          @assert temp_mat[1, 1] == identity
+          @assert temp_mat[3, 3] == identity
+          Hmat[range_H + 1 - n, range_H - n] = temp_mat[2, 2]
+          Hmat[end, range_H - n] = temp_mat[3, 2]
+          Hmat[range_H + 1 - n, 1] = temp_mat[2, 1]
+          Hmat[range_H + 1 - n, range_H + 1 - n] *= ITensor(
+            T, filterinds(commoninds(temp_mat[2, 2], temp_mat[2, 1]); tags="Link")
```

Comment:
Do we need the tag filter here? Best to try to avoid that.

Reply:
I removed it. The lines were not strictly necessary anyway.